File: android/app/src/main/java/com/iitcalendar/applet/DailyTimesWidgetProvider.java
package com.iitcalendar.applet;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import android.widget.RemoteViews;

import org.json.JSONArray;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class DailyTimesWidgetProvider extends AppWidgetProvider {
    private static final String TAG = "DailyTimesWidget";

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        Log.d(TAG, "onUpdate called for " + appWidgetIds.length + " widgets");
        try {
            SharedPreferences prefs = context.getSharedPreferences("group.iit.calendar", Context.MODE_PRIVATE);
            String sunTimesRaw = prefs.getString("sun_times", "[]");
            if (sunTimesRaw == null) sunTimesRaw = "[]";
            Log.d(TAG, "Raw sun times: " + sunTimesRaw);

            String dawnText = "--:--";
            String noonText = "--:--";

            try {
                JSONArray sunArray = new JSONArray(sunTimesRaw);
                if (sunArray.length() > 0) {
                    // Find today's entry
                    String todayStr = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(new Date());
                    Log.d(TAG, "Today's date string: " + todayStr);
                    boolean found = false;
                    for (int i = 0; i < sunArray.length(); i++) {
                        JSONObject entry = sunArray.getJSONObject(i);
                        if (todayStr.equals(entry.optString("date"))) {
                            dawnText = entry.optString("dawn", "--:--");
                            noonText = entry.optString("noon", "--:--");
                            found = true;
                            Log.d(TAG, "Found today's entry: dawn=" + dawnText + ", noon=" + noonText);
                            break;
                        }
                    }
                    // Fallback to first entry if today not found
                    if (!found) {
                        Log.w(TAG, "Today's entry not found, falling back to first entry");
                        JSONObject first = sunArray.getJSONObject(0);
                        dawnText = first.optString("dawn", "--:--");
                        noonText = first.optString("noon", "--:--");
                    }
                } else {
                    Log.w(TAG, "sunArray is empty");
                }
            } catch (Exception e) {
                Log.e(TAG, "Error parsing JSON array", e);
            }

            for (int appWidgetId : appWidgetIds) {
                Log.d(TAG, "Updating widget id: " + appWidgetId);
                RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_daily_times);
                views.setTextViewText(R.id.tvDawnTime, dawnText);
                views.setTextViewText(R.id.tvNoonTime, noonText);
                appWidgetManager.updateAppWidget(appWidgetId, views);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error in onUpdate", e);
        }
    }
}

File: android/app/src/main/java/com/iitcalendar/applet/DailyTimesWidgetProvider.kt
package com.iitcalendar.applet

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.widget.RemoteViews
import org.json.JSONArray
import java.text.SimpleDateFormat
import java.util.*

class DailyTimesWidgetProvider : AppWidgetProvider() {
    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        val prefs = context.getSharedPreferences("group.iit.calendar", Context.MODE_PRIVATE)
        val sunTimesRaw = prefs.getString("sun_times", "[]") ?: "[]"
        
        var dawnText = "--:--"
        var noonText = "--:--"
        
        try {
            val sunArray = JSONArray(sunTimesRaw)
            if (sunArray.length() > 0) {
                // Find today's entry
                val todayStr = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())
                var found = false
                for (i in 0 until sunArray.length()) {
                    val entry = sunArray.getJSONObject(i)
                    if (entry.optString("date") == todayStr) {
                        dawnText = entry.optString("dawn", "--:--")
                        noonText = entry.optString("noon", "--:--")
                        found = true
                        break
                    }
                }
                // Fallback to first entry if today not found
                if (!found) {
                    val first = sunArray.getJSONObject(0)
                    dawnText = first.optString("dawn", "--:--")
                    noonText = first.optString("noon", "--:--")
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        for (appWidgetId in appWidgetIds) {
            val views = RemoteViews(context.packageName, R.layout.widget_daily_times)
            views.setTextViewText(R.id.tvDawnTime, dawnText)
            views.setTextViewText(R.id.tvNoonTime, noonText)
            appWidgetManager.updateAppWidget(appWidgetId, views)
        }
    }
}
